### ISSUE TITLE
`toString` of byte arrays should be represented as base64/hex

### ISSUE THREAD

**Issue by felixdesouza** [immutables/immutables#758]
If I have an example like this: 

```
@Value.Immutable
@JsonDeserialize(as = ImmutableExample.class)
interface Example {
  byte[] contents();
}
```

It works pretty nicely with jackson since it automatically base64 encodes/decodes and we don't have to handle that logic ourselves. However, invoking `toString()` (commonly via logging) yields a quite unfavourable representation of said bytes.

e.g.:
```
Example{contents=[10, 37, 10, 31, 10, 9, 109, 117, 108, 116, 105, 112, 97, 115, 115, 18, 18, 112, 114, 101,
 114, 101, 103, 105, 115, 116, 101, 114, 101, 100, 85, 115, 101, 114, 115, 18, 2, 10, 0, 10, 33, 10, 27, 10, 9,
 109, 117, 108, 116, 105, 112, 97, 115, 115, 18, 14, 114, 101, 97, 108, 109, 73, 110, 100, 101, 120, 95, 105,
 100, 120, 18, 2, 10, 0, 10, 54, 10, 48, 10, 9, 109, 117, 108, 116, 105, 112, 97, 115, 115, 18, 35, 112, 114,
 101, 114, 101, 103, 105, 115, 116, 101, 114, 101, 100, 85, 115, 101, 114, 115, 66, 121, 80, 114, 105, 110,
 99, 105, 112, 97, 108, 73, 100, 95, 105, 100, 120, 18, 2, 10, 0, 10, 31, 10, 25, 10, 9, 109, 117, 108, 116,
 105, 112, 97, 115, 115, 18, 12, 97, 112, 105, 77, 105, 103, 114, 97, 116, 105, 111, 110, 18, 2, 10, 0, 10, 33,
 10, 27, 10, 9, 109, 117, 108, 116, 105, 112, 97, 115, 115, 18, 14, 112, 114, 101, 112, 97, 114, 101, 100, 69,
 118, 101, 110, 116, 115, 18, 2, 10, 0, 10, 44, 10, 38, 10, 9, 109, 117, 108, 116, 105, 112, 97, 115, 115, 18,
 25, 117, 115, 101, 114, 110, 97, 109, 101, 83, 104, 97, 82, 101, 97, 108, 109, 73, 110, 100, 101, 120, 95,
 105, 100, 120, 18, 2, 10, 0, 10, 46, 10, 40, 10, 9, 109, 117, 108, 116, 105, 112, 97, 115, 115, 18, 27, 105,
 110, 100, 101, 120, 67, 111, 110, 115, 117, 109, 101, 114, 83, 101, 113, 117, 101, 110, 99, 101, 80, 111, 105,
 110, 116, 115, 18, 2, 10, 0, 10, 40, 10, 34, 10, 9, 109, 117, 108, 116, 105, 112, 97, ...]}
```

I would propose having the default be base64, as that is slightly more legible than a large array of numbers. I believe this is also safe to do by default as it seems unlikely/unreasonable/strange for tests to assert the contents of a `toString()` especially from a generated class.

If people do desire the old behaviour, it could be opened up via a `@Value.Style` annotation? Only downside is that it would apply to all properties within the class, unless you can apply `@Value.Style` on individual properties?

Thoughts? Ideas?

**Comment by felixdesouza** [comment]
@elucash any update on this by any change? I'm happy to have a stab at it myself, although I would need a couple pointers where to start 😅

**Comment by elucash** [comment]
@felixdesouza thank you for reminding. I think that default implementation can be changed. First of all, It would make sense to somehow limit the max length of any output. On the other hand, it may be an unnecessary complication. For byte array I somehow feel that space-separated printout of hex values might make a good default (`Example{contents=FA 32 41...`), this is not solving the general problem of any custom formatting necessary. As a practical solution, I would highly recommend one of the following:

* Wrapper object for `toString`: http://immutables.github.io/immutable.html#opaque-containers
* Create custom encoding and specify `toString`: http://immutables.github.io/encoding.html

The default for byte array can be fixed, but it may not satisfy all use cases. Maybe even suppressed for all arrays, like `Exampe{contents=<byte[237]>`... specifying only type and length

**Comment by felixdesouza** [comment]
Oop, terribly sorry for the late reply. I suppose I was just looking for a sensible default, it comes from internal libraries published by other teams. It would be unlikely without enforcement that they would opt in/remember to opt in to have opaque containers/custom encoders. Just being able to bump version would be an easier goal, since not everyone cares either ha 😜 😢

**Comment by elucash** [comment]
@felixdesouza also, please, do not overlook the possibility to use an encoding http://immutables.github.io/encoding.html which can be enforced on a package/parent packet (at least in java8) or on a style mete-annotation if you use one.

**Comment by felixdesouza** [comment]
Ooh very interesting, I'd read it before, but I hadn't made the link. A question though, the moment I make a byte array encoding, do I lose the nice byte array methods (although there may not be much, so it's probably minor) that are usually generated for me? 

I suppose it's better phrased as, do encodings act as overrides or can I pull in defaults that can come about from either the existing version (current helper conveniences) or a later version (new helper conveniences)?

I'll have a go at making my byte array encoding, but I would be interested in the above questions.

On a somewhat unrelated, but meta thought: after releasing encodings, did the thought of converting all default encodings to use the Encoder APIs cross your mind? I haven't dug deep in the code, but thought it'd be quicker to ask.

**Comment by elucash** [comment]
Yes, encoding is a full override, but you can encode all convenience methods you need in the way you want, for example, avoid copying arrays etc or add more handy shortcuts.

For the next major iteration of _Immutables_ (if there will be such), all encodings will be externalized and pluggable separately, this would make the most sense, of course

**Comment by elucash** [comment]
under additional review, I think that there's no need to change the default because every special form can be found controversial and not worth to introduce custom routines into generated code besides standard `Arrays.toString` calls. I highly recommend using custom wrapper objects or encodings to handle such cases (as mentioned above).
